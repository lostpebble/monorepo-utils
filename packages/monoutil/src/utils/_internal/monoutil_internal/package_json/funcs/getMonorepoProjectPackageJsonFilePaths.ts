import { glob } from "tinyglobby";

export async function getMonorepoProjectPackageJsonFilePaths(): Promise<string[]> {
  return await glob(["**/package.json", "!**/node_modules/**"]);
}
