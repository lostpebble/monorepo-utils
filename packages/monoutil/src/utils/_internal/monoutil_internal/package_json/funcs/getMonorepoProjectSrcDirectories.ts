import { glob } from "tinyglobby";

export async function getMonorepoProjectSrcDirectories(): Promise<string[]> {
  return (
    await glob(["**/src", "!**/node_modules/**"], {
      onlyDirectories: true,
    })
  ).filter((dir) => dir.endsWith("src/"));
}
