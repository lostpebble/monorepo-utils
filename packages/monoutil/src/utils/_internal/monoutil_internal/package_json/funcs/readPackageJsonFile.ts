import { readJsonFile } from "../../files";
import { IPackageJsonFile } from "../package_json.types";

export async function readPackageJsonFile(filePath: string): Promise<IPackageJsonFile> {
  return readJsonFile(filePath, "module package.json");
}
