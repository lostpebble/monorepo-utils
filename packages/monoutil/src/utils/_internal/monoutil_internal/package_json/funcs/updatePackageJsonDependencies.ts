import { EPackageDependencyType } from "../package_json.enums";
import { ALL_PACKAGE_DEPENDENCY_TYPES, DEPENDENCY_TYPE_TO_KEY } from "../package_json.static";
import { IPackageDependencyUpdate, IPackageJsonFile, IUpdatedDep } from "../package_json.types";

interface IUpdatePackageDependencies_Input<P extends IPackageJsonFile> {
  packageJson: P;
  dependencyUpdates: IPackageDependencyUpdate[];
}

interface IUpdatePackageDependencies_Output<P extends IPackageJsonFile> {
  newPackageJson: P;
  updatedDeps: IUpdatedDep[];
}

export function updatePackageJsonDependencies<P extends IPackageJsonFile>({
  packageJson,
  dependencyUpdates,
}: IUpdatePackageDependencies_Input<P>): IUpdatePackageDependencies_Output<P> {
  const newPackageJson: P = {
    ...packageJson,
  };

  const updatedDeps: IUpdatedDep[] = [];

  function updateDependency(type: EPackageDependencyType, depUpdate: IPackageDependencyUpdate) {
    if (
      newPackageJson[DEPENDENCY_TYPE_TO_KEY[type]]?.[depUpdate.name] != null &&
      newPackageJson[DEPENDENCY_TYPE_TO_KEY[type]]![depUpdate.name] !== depUpdate.version
    ) {
      const previousVersion = newPackageJson[DEPENDENCY_TYPE_TO_KEY[type]]![depUpdate.name];
      newPackageJson[DEPENDENCY_TYPE_TO_KEY[type]]![depUpdate.name] = depUpdate.version;

      updatedDeps.push({
        name: depUpdate.name,
        version: depUpdate.version,
        type,
        previousVersion,
      });
    }
  }

  for (const dep of dependencyUpdates) {
    const updateTypes = dep.updateTypes ?? ALL_PACKAGE_DEPENDENCY_TYPES;

    const updateProd = updateTypes.includes(EPackageDependencyType.production);
    const updateDev = updateTypes.includes(EPackageDependencyType.dev);
    const updatePeer = updateTypes.includes(EPackageDependencyType.peer);
    const updateOverrides = updateTypes.includes(EPackageDependencyType.override);
    const updateResolutions = updateTypes.includes(EPackageDependencyType.resolution);

    if (updateProd) {
      updateDependency(EPackageDependencyType.production, dep);
    }

    if (updateDev) {
      updateDependency(EPackageDependencyType.dev, dep);
    }

    if (updatePeer) {
      updateDependency(EPackageDependencyType.peer, dep);
    }

    if (updateResolutions) {
      updateDependency(EPackageDependencyType.resolution, dep);
    }

    if (updateOverrides) {
      updateDependency(EPackageDependencyType.override, dep);
    }
  }

  return {
    newPackageJson,
    updatedDeps,
  };
}
