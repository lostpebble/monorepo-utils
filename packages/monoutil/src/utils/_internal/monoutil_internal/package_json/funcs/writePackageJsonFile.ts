import { writeJsonFile } from "../../files";
import { IPackageJsonFile } from "../package_json.types";

export async function writePackageJsonFile<P extends IPackageJsonFile>(
  filePath: string,
  packageJson: P,
): Promise<void> {
  await writeJsonFile({
    json: packageJson,
    filePurpose: "module package.json",
    filePath,
    shouldExist: true,
  });
}
