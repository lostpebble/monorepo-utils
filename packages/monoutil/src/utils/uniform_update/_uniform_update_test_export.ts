import * as enums from "./uniform_update.enums";
import * as funcs from "./uniform_update.func";
import * as zods from "./uniform_update.zod";

export const _uniformUpdateTestExport = {
  funcs,
  zods,
  enums,
};
