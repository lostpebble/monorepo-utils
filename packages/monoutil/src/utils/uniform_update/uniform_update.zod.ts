import { z } from "zod";
import { EPackageDependencyType } from "../_internal/monoutil_internal/package_json/package_json.enums";

export const zTargetDependencyTypes = z.object({
  exclude: z.array(z.nativeEnum(EPackageDependencyType)).optional(),
  include: z.array(z.nativeEnum(EPackageDependencyType)).optional(),
});

export const zUniformUpdateTargetVersion = z.object({
  version: z.string(),
  dependencies: z.array(z.string()),
  dependencyTypes: zTargetDependencyTypes.optional(),
});

export const zUniformUpdateConfig = z.object({
  targetVersions: z.array(zUniformUpdateTargetVersion).optional(),
  dependencyTypes: zTargetDependencyTypes.optional(),
});

export type TUniformUpdateConfig = z.infer<typeof zUniformUpdateConfig>;
export type TUniformUpdateTargetVersion = z.infer<typeof zUniformUpdateTargetVersion>;
