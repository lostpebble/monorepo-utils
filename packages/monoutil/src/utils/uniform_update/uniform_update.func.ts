import { $ } from "bun";
import { EMonoutilId } from "../../monoutil.types";
import { readJsonFile } from "../_internal/monoutil_internal/files";
import { createLogger } from "../_internal/monoutil_internal/logging";
import { getMonorepoProjectPackageJsonFilePaths } from "../_internal/monoutil_internal/package_json/funcs/getMonorepoProjectPackageJsonFilePaths";
import { readPackageJsonFile } from "../_internal/monoutil_internal/package_json/funcs/readPackageJsonFile";
import { updatePackageJsonDependencies } from "../_internal/monoutil_internal/package_json/funcs/updatePackageJsonDependencies";
import { writePackageJsonFile } from "../_internal/monoutil_internal/package_json/funcs/writePackageJsonFile";
import { EPackageDependencyType } from "../_internal/monoutil_internal/package_json/package_json.enums";
import { ALL_PACKAGE_DEPENDENCY_TYPES } from "../_internal/monoutil_internal/package_json/package_json.static";
import {
  IPackageDependencyUpdate,
  IUpdatedDep,
} from "../_internal/monoutil_internal/package_json/package_json.types";
import {
  TUniformUpdateConfig,
  TUniformUpdateDependencyTypes,
  zUniformUpdateConfig,
} from "./uniform_update.zod";

const uniformLogger = createLogger(EMonoutilId.uniform_update);

function updateDependencyObject(
  dependencies: Record<string, string>,
  depsToUpdate: string[],
  newVersion: string,
): Record<string, string> {
  for (const [depName] of Object.entries(dependencies)) {
    if (depsToUpdate.includes(depName)) {
      dependencies[depName] = newVersion;
    }
  }

  return dependencies;
}

export async function getUniformUpdateConfig(
  configFilePath?: string,
): Promise<TUniformUpdateConfig> {
  const filePath = configFilePath ?? "./monoutil/uniform_update.config.json";
  uniformLogger.info(`Loading config from ${filePath}`);

  const configJson = await readJsonFile(filePath, "uniform update config");

  return zUniformUpdateConfig.parse(configJson);
}

export async function uniformUpdate(config: TUniformUpdateConfig): Promise<void> {
  uniformLogger.info("Running with config", config);

  const monorepoPackages = await getMonorepoProjectPackageJsonFilePaths();

  uniformLogger.info(
    `Found ${monorepoPackages.length} package.json files in monorepo:\n  - ${monorepoPackages.join("\n  - ")}`,
  );

  if (config.targetVersions) {
    for (const targetVersion of config.targetVersions) {
      // console.log(`Updating to version: ${config.targetVersions}`);
      uniformLogger.info(
        `Updating to dependency versions:\n  - ${targetVersion.dependencies.map((name) => `"${name}": "${targetVersion.version}"`).join("\n  - ")}`,
      );
    }
  }

  for (const monoPackage of monorepoPackages) {
    const packageJson = await readPackageJsonFile(monoPackage);

    let latestPackageJson = packageJson;
    const updatedDeps: IUpdatedDep[] = [];
    const packageUpdates: IPackageDependencyUpdate[] = [];

    if (config.targetVersions) {
      for (const targetVersion of config.targetVersions) {
        // console.log(`Updating to version: ${config.targetVersions}`);
        // uniformLogger.info(
        //   `Updating to version: "${targetVersion.version}" for deps:\n  - ${targetVersion.dependencies.join("\n  - ")}`,
        // );
        const updateTypes = getDependencyTypes(
          config.dependencyTypes,
          targetVersion.dependencyTypes,
        );

        // uniformLogger.info(
        //   `Attempting update of dependencies ["${targetVersion.dependencies.join(`", "`)}"] to version "${targetVersion.version}" in "./${monoPackage}" for package dependency types: ${updateTypes.join(", ")}`,
        // );
        packageUpdates.push(
          ...targetVersion.dependencies.map(
            (depName): IPackageDependencyUpdate => ({
              updateTypes,
              version: targetVersion.version,
              name: depName,
            }),
          ),
        );

        /*const update = updatePackageJsonDependencies({
          packageJson,
          dependencyUpdates: targetVersion.dependencies.map(
            (depName): IPackageDependencyUpdate => ({
              updateTypes,
              version: targetVersion.version,
              name: depName,
            }),
          ),
        });*/
      }
    }

    if (config.targetDependencies != null) {
      for (const targetDependencyObject of config.targetDependencies) {
        const updateTypes = getDependencyTypes(
          config.dependencyTypes,
          targetDependencyObject.dependencyTypes,
        );

        const depNames = Object.keys(targetDependencyObject.dependencies);

        // uniformLogger.info(
        //   `Updating dependencies:\n  - ${depNames.map((depName) => `"${depName}": ${targetDependencyObject.dependencies[depName]}`).join("\n  - ")}`,
        // );

        const packageUpdate = depNames.map((depName): IPackageDependencyUpdate => {
          return {
            updateTypes,
            version: targetDependencyObject.dependencies[depName],
            name: depName,
          };
        });

        packageUpdates.push(...packageUpdate);
      }
    }

    const update = updatePackageJsonDependencies({
      packageJson,
      dependencyUpdates: packageUpdates,
    });

    if (update.updatedDeps.length > 0) {
      latestPackageJson = update.newPackageJson;
      updatedDeps.push(...update.updatedDeps);
    }

    if (latestPackageJson !== packageJson) {
      await writePackageJsonFile(monoPackage, latestPackageJson);
      await $`bunx biome format --write ${monoPackage}`;

      if (updatedDeps.length > 0) {
        uniformLogger.log(
          `Updated dependencies in "./${monoPackage}":\n  - ${updatedDeps.map((updatedDep) => `(${updatedDep.type} dependency) [${updatedDep.name}] "${updatedDep.version}" (was "${updatedDep.previousVersion}")`).join("\n  - ")}`,
        );
      }
    }
  }
}

function getDependencyTypes(
  ...depTypes: (TUniformUpdateDependencyTypes | undefined)[]
): EPackageDependencyType[] {
  const mergedIncludes = [...new Set(depTypes.flatMap((d) => d?.include))].filter((d) => d != null);
  const mergedExcludes = [...new Set(depTypes.flatMap((d) => d?.exclude))].filter((d) => d != null);

  return (mergedIncludes.length === 0 ? [...ALL_PACKAGE_DEPENDENCY_TYPES] : mergedIncludes).filter(
    (dep) => !mergedExcludes.includes(dep),
  );
}
