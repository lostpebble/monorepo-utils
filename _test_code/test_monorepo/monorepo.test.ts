import { afterAll, beforeEach, describe, expect, it } from "bun:test";
import { $ } from "bun";
import { _uniformUpdateTestExport } from "../../packages/monoutil/src/utils/uniform_update/_uniform_update_test_export";
import { getTestProjectPackageJsonObject } from "./packages/test_src/funcs/getTestProjectPackageJson";

const { zods, funcs, enums } = _uniformUpdateTestExport;
const nothing = { zods, funcs, enums };

const originalCwd = process.cwd();

describe("monorepo utils", () => {
  beforeEach(async () => {
    process.chdir(__dirname);
    $.cwd(__dirname);
    await $`bun run reset_test.ts`;
  });

  afterAll(() => {
    $.cwd(originalCwd);
    process.chdir(originalCwd);
  });

  describe("uniform_update", () => {
    it("should update the monorepo dependencies with a passed CLI arguments", async () => {
      await $`bun run monoutil_bin.ts uniform_update --module pullstate --version 2.0.0`;
      const testPackageJson = await getTestProjectPackageJsonObject();
      expect(testPackageJson.dependencies).not.toBeUndefined();
      expect(testPackageJson.dependencies!["pullstate"]).not.toBeUndefined();
      expect(testPackageJson.dependencies!["pullstate"]!).toEqual("2.0.0");
    });
  });
});
