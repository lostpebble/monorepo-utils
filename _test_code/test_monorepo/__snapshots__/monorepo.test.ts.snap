// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`monorepo utils uniform_update should update the monorepo dependencies with a config file 1`] = `
"{
  "name": "@monoutil/test_project_package",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {},
  "dependencies": {
    "pullstate": "2.0.0",
    "rimraf": "6.0.1",
    "lodash": "4.17.21"
  },
  "devDependencies": { "monoutil": "workspace:*" },
  "repository": {
    "type": "git",
    "url": "https://github.com/lostpebble/monorepo-utils.git"
  },
  "private": true
}
"
`;

exports[`monorepo utils uniform_update should update the monorepo dependencies with a passed CLI arguments 1`] = `
"{
  "name": "@monoutil/test_project_package",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {},
  "dependencies": {
    "pullstate": "2.0.0",
    "rimraf": "6.0.1",
    "lodash": "4.17.21"
  },
  "devDependencies": { "monoutil": "workspace:*" },
  "repository": {
    "type": "git",
    "url": "https://github.com/lostpebble/monorepo-utils.git"
  },
  "private": true
}
"
`;

exports[`monorepo utils uniform_update should update the monorepo dependencies with a passed CLI arguments 2`] = `
"{
  "name": "@monoutil/test_project_package",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {},
  "dependencies": {
    "pullstate": "2.0.0",
    "rimraf": "6.0.1",
    "lodash": "4.17.21"
  },
  "devDependencies": { "monoutil": "workspace:*" },
  "repository": {
    "type": "git",
    "url": "https://github.com/lostpebble/monorepo-utils.git"
  },
  "private": true
}
"
`;

exports[`monorepo utils uniform_update should update the monorepo dependencies with a config file 2`] = `
"{
  "name": "@monoutil/test_project_package",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {},
  "dependencies": {
    "pullstate": "2.0.0",
    "rimraf": "6.0.1",
    "lodash": "4.17.21"
  },
  "devDependencies": { "monoutil": "workspace:*" },
  "repository": {
    "type": "git",
    "url": "https://github.com/lostpebble/monorepo-utils.git"
  },
  "private": true
}
"
`;
